dre-service/src/main/java/io/cdap/re/YARERulebookHandler.java

```diff
@@ -94,303 +87,380 @@ public void validateWhen(HttpServiceRequest request, HttpServiceResponder respon
 
   @POST
   @Path("rules")
+  @TransactionPolicy(value = TransactionControl.EXPLICIT)
   public void create(HttpServiceRequest request, HttpServiceResponder responder) {
-    try {
-      RequestExtractor handler = new RequestExtractor(request);
-      String content = handler.getContent(StandardCharsets.UTF_8);
-      RuleRequest rule = gson.fromJson(content, RuleRequest.class);
-      rulesDB.createRule(rule);
-
-      JsonObject response = new JsonObject();
-      response.addProperty("status", HttpURLConnection.HTTP_OK);
-      response.addProperty("message", String.format("Successfully created rule '%s'.", rule.getId()));
-      response.addProperty("count", 1);
-      JsonArray values = new JsonArray();
-      values.add(new JsonPrimitive(rule.getId()));
-      response.add("values", values);
-      ServiceUtils.sendJson(responder, HttpURLConnection.HTTP_OK, response.toString());
-    } catch (RuleAlreadyExistsException e) {
-      ServiceUtils.error(responder, HttpURLConnection.HTTP_NOT_FOUND, e.getMessage());
-    } catch (Exception e) {
-      ServiceUtils.error(responder, HttpURLConnection.HTTP_INTERNAL_ERROR,
-                         String.format("Unexpected error while creating rule. Please check your request. %s",
-                                       e.getMessage())
-      );
-    }
+    TransactionRunners.run(getContext(), context -> {
+      try {
+        RequestExtractor handler = new RequestExtractor(request);
+        String content = handler.getContent(StandardCharsets.UTF_8);
+        RuleRequest rule = gson.fromJson(content, RuleRequest.class);
+        RulesDB rulesDB = RulesDB.get(context);
+        rulesDB.createRule(rule);
+
+        JsonObject response = new JsonObject();
+        response.addProperty("status", HttpURLConnection.HTTP_OK);
+        response.addProperty("message", String.format("Successfully created rule '%s'.", rule.getId()));
+        response.addProperty("count", 1);
+
+        JsonArray values = new JsonArray();
+        values.add(new JsonPrimitive(rule.getId()));
+
+        response.add("values", values);
+
+        ServiceUtils.sendJson(responder, HttpURLConnection.HTTP_OK, response.toString());
+      } catch (RuleAlreadyExistsException e) {
+        ServiceUtils.error(responder, HttpURLConnection.HTTP_NOT_FOUND, e.getMessage());
+      } catch (Exception e) {
```

Comment:
this is poor error handling. If the request is invalid, this endpoint will return a 500 instead of 400. Please open a jira to fix the error handling in this app.

Reply:
Will do

Reply:
JIRA: https://issues.cask.co/browse/CDAP-15697